File: backend/main/java/de/griefed/serverpackcreator/versionmeta/legacyfabric/LegacyFabricVersioning.java
/* Copyright (C) 2022  Griefed
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301
 * USA
 *
 * The full license can be found at https:github.com/Griefed/ServerPackCreator/blob/main/LICENSE
 */
package de.griefed.serverpackcreator.versionmeta.legacyfabric;

import de.griefed.serverpackcreator.utilities.common.Utilities;
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import org.jetbrains.annotations.NotNull;

/**
 * LegacyFabric version parent-class, implemented by {@link LegacyFabricGame} and
 * {@link LegacyFabricLoader} to store and provide version information for game and loader
 * versions.
 */
abstract class LegacyFabricVersioning {

  private final List<String> RELEASES = new ArrayList<>(100);
  private final List<String> SNAPSHOTS = new ArrayList<>(100);
  private final List<String> ALL = new ArrayList<>(200);
  private final Utilities UTILITIES;
  private final File MANIFEST;

  /**
   * @param manifest  The manifest holding the version information for this LegacyFabric type.
   * @param utilities Commonly used utilities across ServerPackCreator.
   */
  LegacyFabricVersioning(@NotNull File manifest,
                         @NotNull Utilities utilities) {
    UTILITIES = utilities;
    MANIFEST = manifest;
  }

  /**
   * Update all lists of available versions with new information gathered from the manifest.
   *
   * @throws IOException When the manifest could not be read.
   */
  void update() throws IOException {
    RELEASES.clear();
    SNAPSHOTS.clear();
    ALL.clear();

    UTILITIES.JsonUtilities().getJson(MANIFEST).forEach(node -> {
      String version = node.get("version").asText();
      ALL.add(version);
      if (node.get("stable").asBoolean()) {
        RELEASES.add(version);
      } else {
        SNAPSHOTS.add(version);
      }
    });
  }

  /**
   * List of release versions for this meta.
   *
   * @return Release versions.
   * @author Griefed
   */
  @NotNull List<String> releases() {
    return RELEASES;
  }

  /**
   * List of snapshot/pre-release versions for this meta.
   *
   * @return Snapshot/pre-release versions.
   * @author Griefed
   */
  @NotNull List<String> snapshots() {
    return SNAPSHOTS;
  }

  /**
   * List of all versions for this meta.
   *
   * @return All versions.
   * @author Griefed
   */
  @NotNull List<String> all() {
    return ALL;
  }

}

File: serverpackcreator-api/src/main/kotlin/de/griefed/serverpackcreator/api/versionmeta/legacyfabric/LegacyFabricVersioning.kt
/* Copyright (C) 2026 Griefed
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301
 * USA
 *
 * The full license can be found at https:github.com/Griefed/ServerPackCreator/blob/main/LICENSE
 */
package de.griefed.serverpackcreator.api.versionmeta.legacyfabric

import de.griefed.serverpackcreator.api.utilities.common.Utilities
import de.griefed.serverpackcreator.api.versionmeta.VersionMetaConfig
import java.io.File
import java.io.IOException
import java.util.Collections

/**
 * LegacyFabric version parent-class, implemented by [LegacyFabricGame] and
 * [LegacyFabricLoader] to store and provide version information for game and loader
 * versions.
 *
 * @param manifest  The manifest holding the version information for this LegacyFabric type.
 * @param utilities Commonly used utilities across ServerPackCreator.
 * @author Griefed
 */
internal abstract class LegacyFabricVersioning(
    private val manifest: File,
    private val utilities: Utilities
) {
    /**
     * Published as an **immutable snapshot behind `@Volatile`**, not as a collection [update] mutates in
     * place. The refresh runs on a background coroutine while callers read; clearing and refilling a
     * shared list let a reader throw `ConcurrentModificationException` or silently observe the empty
     * window between the two.
     */
    @Volatile
    var releases: List<String> = emptyList()
        private set
    /**
     * Published as an **immutable snapshot behind `@Volatile`**, not as a collection [update] mutates in
     * place. The refresh runs on a background coroutine while callers read; clearing and refilling a
     * shared list let a reader throw `ConcurrentModificationException` or silently observe the empty
     * window between the two.
     */
    @Volatile
    var snapshots: List<String> = emptyList()
        private set
    /**
     * Published as an **immutable snapshot behind `@Volatile`**, not as a collection [update] mutates in
     * place. The refresh runs on a background coroutine while callers read; clearing and refilling a
     * shared list let a reader throw `ConcurrentModificationException` or silently observe the empty
     * window between the two.
     */
    @Volatile
    var allVersions: List<String> = emptyList()
        private set

    /**
     * Update all lists of available versions with new information gathered from the manifest.
     *
     * @throws IOException When the manifest could not be read.
     * @author Griefed
     */
    @Throws(IOException::class)
    fun update() {
        val next_releases = ArrayList<String>(100)
        val next_snapshots = ArrayList<String>(100)
        val next_allVersions = ArrayList<String>(200)
        for (node in utilities.jsonUtilities.getJson(manifest)) {
            val version: String = node.get(VersionMetaConfig.TAG_VERSION).asText()
            val stable = node.get(VersionMetaConfig.TAG_STABLE).asBoolean()
            next_allVersions.add(version)
            if (stable) {
                next_releases.add(version)
            } else {
                next_snapshots.add(version)
            }
        }
            // Published in one assignment each, as unmodifiable views: a `List`-typed field still
        // holds an ArrayList at runtime, so a caller could otherwise cast and mutate our state.
        releases = Collections.unmodifiableList(next_releases)
        snapshots = Collections.unmodifiableList(next_snapshots)
        allVersions = Collections.unmodifiableList(next_allVersions)
}
}
